[Zhi Hong] iP

ongzhihong opened the pull request.

ongzhihong committed:
Add Increment Level-0 (Greet)

ongzhihong committed:
Add Increment Level-1 Greet Echo Exit

ongzhihong committed:
Add Increment Level-2 Add List

ongzhihong committed:
Add Increment Level-3 Mark As Done

Refactor Duke class to move methods into DukeController class
Add Task class

ongzhihong committed:
Fix code to comply with coding standard

yfg-dev commented on src/main/java/DukeController.java:
```diff
@@ -0,0 +1,68 @@
+import java.util.ArrayList;
+import java.util.Scanner;
```
Good job for listing explicitly

yfg-dev commented on src/main/java/Duke.java:
```diff
@@ -1,10 +1,24 @@
-                + "| |_| | |_| |   <  __/\n"
-                + "|____/ \\__,_|_|\\_\\___|\n";
-        System.out.println("Hello from\n" + logo);
+        DukeController dukeController = new DukeController();
```
Good use of naming for class

yfg-dev commented on src/main/java/Task.java:
```diff
@@ -0,0 +1,19 @@
+        return (isDone ? "X" : " "); // mark done task with X
+    }
+
+    public void updateStatus(boolean status){
```
Clear naming

yfg-dev commented on src/main/java/DukeController.java:
```diff
@@ -0,0 +1,68 @@
+import java.util.ArrayList;
+import java.util.Scanner;
+
+public class DukeController {
```
Good naming of methods, able to tell its use

yfg25 reviewed (commented):
Good coding standards overall

ongzhihong committed:
Add Level 4 Increment ToDo Deadline Event

wangtingjia commented on src/main/java/Deadline.java:
```diff
@@ -0,0 +1,24 @@
+public class Deadline extends Task {
+
+    private String by;
```
Consider using final for variable by

wangtingjia commented on src/main/java/DukeController.java:
```diff
@@ -0,0 +1,92 @@
+import java.util.Scanner;
+
+public class DukeController {
+    protected ArrayList<Task> taskList = new ArrayList<Task>(0);
```
I like the use of ArrayList data structure to store the tasks

wangtingjia commented on src/main/java/DukeController.java:
```diff
@@ -0,0 +1,92 @@
+        printNewLine();
+
+    }
+}
```
I like the very neat organisation of the functions. This allows for easy single level of abstraction in the main function

wangtingjia commented on src/main/java/Event.java:
```diff
@@ -0,0 +1,20 @@
+    public void printItem() {
+        System.out.println("[E]["+this.getStatusIcon()+"] "+this.description+" (at: "+this.timing+")");
+    }
+}
```
Clear and correct naming of all classes, variables and functions as per coding standards.

wangtingjia commented on src/main/java/Event.java:
```diff
@@ -0,0 +1,20 @@
+    public void setTiming(String timing) {
+        this.timing = timing;
+    }
+    @Override
```
Good use of Override method

wangtingjia commented on src/main/java/Event.java:
```diff
@@ -0,0 +1,20 @@
+public class Event extends Task {
+    private String timing;
```
Consider adding "final" to variable timing

wangtingjia reviewed (commented):
Good quality code that closely follows coding standards. Great effort!

ongzhihong committed:
Add Input parser and TextUITesting files

Refactor methods in DukeController and Duke and rename variables to improve code quality
Fix UI alignment issues with printing

ongzhihong committed:
Add Increment Level-5 Class IllegalInputException to handle illegal inputs

Users are able to input illegal inputs and commands that will crash the program or make it behave not as expected.

An IllegalInputException should be thrown whenever the user enters an illegal input because users should not be able to crash the program or cause bugs in the program.

Let's update the methods parsing user inputs to throw errors and the method handling user input to catch IllegalInputExceptions and display them on the terminal to the user.

ongzhihong committed:
Update Manual Testing

ongzhihong committed:
Merge branch 'branch-level-5'

* branch-level-5:
  Update Manual Testing
  Add Increment Level-5 Class IllegalInputException to handle illegal inputs

ongzhihong committed:
Move Duke classes into Duke Package

ongzhihong committed:
Merge branch 'branch-A-Packages'

* branch-A-Packages:
  Move Duke classes into Duke Package

ongzhihong committed:
Add Increment Level-6 Delete

ongzhihong committed:
Add Increment Level-7 Save

ongzhihong committed:
Merge branch 'branch-Level-6'

* branch-Level-6:
  Add Increment Level-6 Delete

ongzhihong committed:
Merge branch 'branch-Level-7'

* branch-Level-7:
  Add Increment Level-7 Save

ongzhihong committed:
Add proper handling for saving

Handling for data directory not existing
Feedbacking on successful saving of tasks
Bugfix for saving and reading taking different fields

richwill28 commented on src/main/java/duke/DukeController.java:
```diff
@@ -0,0 +1,245 @@
+
+public class DukeController {
+    protected ArrayList<Task> taskList = new ArrayList<Task>(0);
+    private final String GREETING_PART_ONE = "Hello! I'm Duke!";
+    private final String GREETING_PART_TWO = "What can I do for you?";
+    private final String GOODBYE = "Bye. Hope to see you again soon!";
+    private final String NUMBER_OF_TASKS_MESSAGE = "You now have %d tasks";
+    private final String TASK_ADDED_SUCCESSFULLY_MESSAGE = "Got it. I've added this task:";
+    private final String EMPTY_TASK_LIST_MESSAGE = "You have no tasks in your list currently!";
+    private final String LISTING_TASKS_MESSAGE = "Here are the tasks in your list:";
+    private final String UPDATE_TASK_SUCCESS = "Nice! I've %sed this task:";
+    private final String ILLEGAL_INDEX_ERROR = ":( %d is out of bounds of the task list";
+    private final String UPDATE_TASK_FAILURE = ":( This task is already %sed";
+    private final String DELETE_TASK_SUCCESS = "Nice! I've deleted this task:";
+    private final String SAVE_FILE_SUCCESS = "Nice! I've saved all your tasks!";
+    private final String LOAD_FILE_SUCCESS = "Nice! I've loaded your tasks from previous sessions!";
+    private final String CREATE_DIRECTORY_FAILED = "Please create an empty directory named %s in the folder containing ur .jar file!";
+    private final int TASK_TYPE = 0;
+    private final int TASK_STATUS = 1;
+    private final int TASK_DESCRIPTION = 2;
+    private final int TASK_TIMING = 3;
+    private final String DATA_DIRECTORY = "data";
```
[Code Quality] Consider isolating these UI-related methods to a dedicated UI class.

richwill28 commented on src/main/java/duke/DukeController.java:
```diff
@@ -0,0 +1,245 @@
+import java.util.Scanner;
+import java.util.regex.Pattern;
+
+public class DukeController {
```
[Code Quality] `DukeController` as a class is probably too broad. Try to figure out how to enforce SRP and better separation of concerns. Also, please add a comment for the class if it's not trivial.

richwill28 commented on src/main/java/duke/DukeController.java:
```diff
@@ -0,0 +1,245 @@
+            case "list":
+                printTaskList();
+                break;
+            case "todo":
+            case "deadline":
+            case "event":
+                try {
+                    String[] details = inputParser.parseTaskInformation(parsedInput[1], userCommand);
+                    Task newTask = generateTask(userCommand, details);
+                    addTask(newTask);
+                    saveData();
+                } catch (IllegalInputException e) {
+                    System.out.println(e.getErrorMessage());
+                    printNewLine();
+                } catch (IOException e){
+                    System.out.println(e.getMessage());
+                    printNewLine();
+                }
+                break;
+            case "delete":
+            case "mark":
+            case "unmark":
+                try {
+                    int taskIndex = inputParser.parseTaskIndex(parsedInput[1], userCommand);
+                    if (userCommand.equals("mark") || userCommand.equals("unmark")) {
+                        updateTaskStatus(taskIndex, userCommand);
+                    } else {
+                        deleteTask(taskIndex);
+                    }
+                    saveData();
+                } catch (IllegalInputException e){
+                    System.out.println(e.getErrorMessage());
+                    printNewLine();
+                } catch (IOException e){
+                    System.out.println(e.getMessage());
+                    printNewLine();
+                }
+                break;
+            default:
+                printNewLine();
```
[Code Quality] The logic inside each case could be extracted out for better SLAP and control flow.

richwill28 commented on src/main/java/duke/DukeController.java:
```diff
@@ -0,0 +1,245 @@
+        printNewLine();
+    }
+
+    public void updateTaskStatus(int index, String userCommand) throws IllegalInputException{
+        if (index > taskList.size() || index < 1){
+            throw new IllegalInputException(String.format(ILLEGAL_INDEX_ERROR,index));
+        }
+        index--;
+        boolean isDone = userCommand.equals("mark");
+        Task task = taskList.get(index);
+        if (isDone == task.isDone()){
+            task.printTask();
+            throw new IllegalInputException(String.format(UPDATE_TASK_FAILURE,userCommand));
+        }
+        task.setDone(isDone);
+        System.out.println(String.format(UPDATE_TASK_SUCCESS, userCommand));
+        task.printTask();
+        printNewLine();
+    }
+
+    public void deleteTask(int index) throws IllegalInputException{
+        if (index > taskList.size() || index < 1){
+            throw new IllegalInputException(String.format(ILLEGAL_INDEX_ERROR,index));
+        }
+        index--;
+        System.out.println(String.format(DELETE_TASK_SUCCESS));
+        taskList.get(index).printTask();
+        taskList.remove(index);
+        printNewLine();
+    };
```
[Code Quality] Task-related methods should probably belong to other classes?

richwill28 reviewed (commented):
Overall, not bad. For the next iterations, you could pay more attention to SRP and also consider how to improve the code architecture.

ongzhihong committed:
Refactor similar code into Storage Ui TaskList classes

ongzhihong committed:
Refactor InputParser class into Parser for standardization

ongzhihong committed:
Add search function for task description

ongzhihong committed:
Merge pull request #1 from ongzhihong/branch-Level-9

Add search function for task description

ongzhihong committed:
Add JavaDoc for most classes and methods

ongzhihong committed:
Merge pull request #2 from ongzhihong/branch-A-JavaDoc

Add JavaDoc for most classes and methods

ongzhihong committed:
Change save file location to be inside current working directory

ongzhihong committed:
Add README
